### ISSUE TITLE
What should the request look like? The response?

### ISSUE THREAD

**Issue by iandees** [RagtagOpen/phone-number-validator#4]
We haven't really gone through the thought process of designing the API yet. Let's figure out if the endpoint I started with fulfills all of the usecases we have in mind.

Right now we have a single endpoint that looks like this:

### `GET v1/number/validate`

[_example_](https://kce1bslyse.execute-api.us-east-1.amazonaws.com/dev/v1/number/validate?number=4158675309)

With query argument `number`, allowing you to specify the number to be validated.

The response looks like this:

```json
{
  "results": [
    {
      "formatted": {
        "e164": "+14158675309", 
        "national": "(415) 867-5309"
      }, 
      "input": "4158675309", 
      "location": {
        "description": "California", 
        "tz": [
          "America/Los_Angeles"
        ]
      }, 
      "reason": "possible", 
      "type": "fixed_or_mobile", 
      "valid": true
    }
  ]
}
```

The root `results` property is an array of results, allowing us to expand to a single request with multiple numbers to check. Each result object has:

Property | Type | Description
---|---|---
`input` | string | The value that was sent in by the user. This can be used to correlate the requested value(s) with the response data.
`valid` | bool | `true` if the number is a parseable/valid US telephone number. `false` if it isn't recognized.
`reason` | string | A short text string describing the `valid` result. (These can be documented later)
`type` | string | A short text string describing the type of phone number if the number is valid.
`formatted` | object | If the number is valid, this will be an object containing the `e164` and `national`-formatted versions of the input number.
`location` | object | If the number is valid, this will be an object containing an English `description` of the (guessed) location for the telephone number and a list of possible timezones (`tz`) that number might sit in.

Does this API fit most of our cases? Anything obvious missing? Any terminology that doesn't make sense?

I'd like to have others think about this before writing tests (#2).

**Comment by iandees** [comment]
I'm not so happy with the method of using multiple query args of the same name to pass in multiple numbers. Technically that's allowed and something like [`?number=200&number=500`](https://kce1bslyse.execute-api.us-east-1.amazonaws.com/dev/v1/number/validate?number=200&number=500) should work, but API Gateway, Lambda, or Zappa seem to munge that and only give the most recent value.

Another option would be to allow comma-separated lists of numbers in a single param, but that makes things complicated when someone accidentally enters a comma in the number they want to validate.

Another option is to not support multiple in this endpoint and require a `POST` with a JSON body or something if you want to do a batch.

**Comment by youknowjack** [comment]
+1 for adding a batch endpoint. It might be worth trying the Facebook API approach, which would look like:
```
curl \
  -F 'batch=[{"method":"GET", "relative_url":"/v1/number/5125551212"},{"method":"GET", "relative_url":"/v1/number/9175551212"}]' \ 
  http://127.0.0.1:5000/batch
```

which would then return a JSON response like:
```
[
    { "code": 200, 
      "body": "{\"formatted\":\"…\"}"},
    { "code": 200,
      "body": "{\"formatted\":\"…\"}"}
]
```

I could put together a PR if you want to see what that code looks like.

**Comment by iandees** [comment]
I like that batch idea, but how does it correlate the request value with the response? Is it only implied by the ordering? If so, I'd prefer to be more explicit than that. It seems like a lot of overhead for folks that just want to throw a bunch of phone numbers at us.

**Comment by youknowjack** [comment]
Pretty sure it's by order, but you could echo back the input in the response.

This alpha project for flask makes the argument for doing batching this way:
https://pypi.org/project/flask-batch/

I see the simplicity argument -- there's not a separate design to be maintained for the batch API. However, in this case I don't think that's a very high cost of maintenance, especially with a little reuse of response format.

In addition, async/await in Python 3 will make it easy to parallelize the implementation so it's not processing one number at a time.

Here's an alternate design. Request:
```
POST /v1/batch
{
  "numbers" : [
    { "number": "512-555-1212" },
    { "number": "1234" },
    ...
  ]
}
```

Response:
- "results" is a JSON object with input numbers as properties
- values reuse /v1/number response schema
```
{
  "results": {
    "512-555-1212": {
      "formatted": {
        "e164": "+15125551212",
        "national": "(512) 555-1212"
      },
      "location": {
        "description": "Texas",
        "tz": [
          "America/Chicago"
        ]
      },
      "reason": "possible",
      "type": "fixed_or_mobile",
      "valid": true
    },
    "1234": {
      "reason": "too_short",
      "valid": false
    },
    ...
  }
}
```

I can create a separate issue and work up a PR for this, if it sounds good.

**Comment by iandees** [comment]
Take a look at #6, which has a similar output to what you suggested, but with an awkward method of submitting multiple numbers. I like your suggestion of POSTing an array of objects. That seems like a reasonable place to go – it gives us somewhere to go if we want to let the user submit more data than just a phone number.

**Comment by Rkauff** [comment]
Hi all, just out of curiosity, has anyone considered using the Twilio Voice AP, and lookup API?They are already built, so we wouldn’t have to reinvent the wheel. 

https://www.twilio.com/docs/voice

-Ryan

**Comment by dryan** [comment]
> Hi all, just out of curiosity, has anyone considered using the Twilio Voice AP, and lookup API?They are already built, so we wouldn’t have to reinvent the wheel.

We have discussed it, but we're hoping to get this available to orgs where budget limitations make the lookup API a bit expensive.

**Comment by iandees** [comment]
I'm going to close this as fixed in #6, but with a huge caveat that I think some piece of our architecture (API Gateway, Zappa, Lambda, etc.) are preventing multiple phone numbers from getting to the Python code.

I'm going to file a new ticket to find a better way to get that data into the API.
